#### Задание 1 ####

Импортируйте библиотеки pandas и numpy. Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn.

In [1]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
boston = load_boston()

In [2]:
boston.data.shape

(506, 13)

Создайте датафреймы X и y из этих данных.

In [3]:
data = boston.data
feature_names = boston.feature_names
target = boston.target

X = pd.DataFrame(data, columns=feature_names)
y = pd.DataFrame(target, columns=['price'])

In [4]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
CRIM       506 non-null float64
ZN         506 non-null float64
INDUS      506 non-null float64
CHAS       506 non-null float64
NOX        506 non-null float64
RM         506 non-null float64
AGE        506 non-null float64
DIS        506 non-null float64
RAD        506 non-null float64
TAX        506 non-null float64
PTRATIO    506 non-null float64
B          506 non-null float64
LSTAT      506 non-null float64
dtypes: float64(13)
memory usage: 51.5 KB


In [5]:
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
price    506 non-null float64
dtypes: float64(1)
memory usage: 4.0 KB


In [6]:
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [7]:
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.3, random_state=42)

In [9]:
X_train.head(5)

CRIM   ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
5     0.02985  0.0   2.18   0.0  0.458  6.430  58.7  6.0622   3.0  222.0   
116   0.13158  0.0  10.01   0.0  0.547  6.176  72.5  2.7301   6.0  432.0   
45    0.17142  0.0   6.91   0.0  0.448  5.682  33.8  5.1004   3.0  233.0   
16    1.05393  0.0   8.14   0.0  0.538  5.935  29.3  4.4986   4.0  307.0   
468  15.57570  0.0  18.10   0.0  0.580  5.926  71.0  2.9084  24.0  666.0   

     PTRATIO       B  LSTAT  
5       18.7  394.12   5.21  
116     17.8  393.30  12.04  
45      17.9  396.90  10.21  
16      21.0  386.85   6.58  
468     20.2  368.74  18.13

In [10]:
X_test.head(5)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
173  0.09178   0.0   4.05   0.0  0.510  6.416  84.1  2.6463   5.0  296.0   
274  0.05644  40.0   6.41   1.0  0.447  6.758  32.9  4.0776   4.0  254.0   
491  0.10574   0.0  27.74   0.0  0.609  5.983  98.8  1.8681   4.0  711.0   
72   0.09164   0.0  10.81   0.0  0.413  6.065   7.8  5.2873   4.0  305.0   
452  5.09017   0.0  18.10   0.0  0.713  6.297  91.8  2.3682  24.0  666.0   

     PTRATIO       B  LSTAT  
173     16.6  395.50   9.04  
274     17.6  396.90   3.53  
491     20.1  390.11  18.07  
72      19.2  390.91   5.52  
452     20.2  385.09  17.27

In [11]:
y_train.head(5)

price
5     28.7
116   21.2
45    19.3
16    23.1
468   19.1

In [12]:
y_test.head(5)

price
173   23.6
274   32.4
491   13.6
72    22.8
452   16.1

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [13]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression

In [14]:
regr = LinearRegression()

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых

In [15]:
regr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [16]:
pred = regr.predict(X_test)

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [17]:
from sklearn.metrics import r2_score

In [18]:
r2_result = r2_score(y_test, pred)
print(r2_result)

0.711226005748496


#### Задание 2 ####

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble. Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42

In [19]:
from sklearn.ensemble import RandomForestRegressor

In [20]:
model = RandomForestRegressor(n_estimators=10000, max_depth=12, random_state=42)

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

In [27]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=12,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10000, n_jobs=None,
           oob_score=False, random_state=42, verbose=0, warm_start=False)

In [28]:
pred = model.predict(X_test)

Сделайте предсказание на тестовых данных и посчитайте R2.

In [29]:
r2_result = r2_score(y_test, pred)
print(r2_result)

0.8730822855114492


Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

**В предыдущем задании результат 0.711226005748496, в этом - 0.8730822855114492. В данном случае модель сработает лучше. Чем больше показатель r2, тем лучше**

#### Задание 3 ####

Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances. 
С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.

In [32]:
print(model.feature_importances_)

[0.03283282 0.001704   0.00713127 0.00109003 0.01377731 0.40598637
 0.01495616 0.06382388 0.00490367 0.01167051 0.01788684 0.01278329
 0.41145386]


In [34]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [35]:
importances = model.feature_importances_
indices = np.argsort(importances)[::-1]
feature_names = boston.feature_names
print("Feature importances:")
for f, idx in enumerate(indices):
    print("{:2d}. feature '{:5s}' ({:.4f})".format(f + 1, feature_names[idx], importances[idx]))

Feature importances:
 1. feature 'LSTAT' (0.4115)
 2. feature 'RM   ' (0.4060)
 3. feature 'DIS  ' (0.0638)
 4. feature 'CRIM ' (0.0328)
 5. feature 'PTRATIO' (0.0179)
 6. feature 'AGE  ' (0.0150)
 7. feature 'NOX  ' (0.0138)
 8. feature 'B    ' (0.0128)
 9. feature 'TAX  ' (0.0117)
10. feature 'INDUS' (0.0071)
11. feature 'RAD  ' (0.0049)
12. feature 'ZN   ' (0.0017)
13. feature 'CHAS ' (0.0011)


Важными показателями являются **LSTAT** и **RM**